File: apps/javascript/design-pattern/tests/proxy.test.js
import { describe, it, expect } from '@jest/globals';
import { BankAccount, protectionProxy, virtualProxy } from '../src/proxy.js';

describe('Proxy パターン', () => {
  describe('BankAccount', () => {
    it('入金・出金ができる', () => {
      const account = new BankAccount('田中', 1000);
      account.deposit(500);
      expect(account.balance).toBe(1500);

      account.withdraw(300);
      expect(account.balance).toBe(1200);
    });

    it('残高不足で出金すると例外が発生する', () => {
      const account = new BankAccount('田中', 100);
      expect(() => account.withdraw(200)).toThrow('残高不足');
    });
  });

  describe('Protection Proxy', () => {
    it('所有者は操作できる', () => {
      const account = new BankAccount('田中', 1000);
      const proxy = protectionProxy(account, '田中');

      proxy.deposit(500);
      expect(proxy.balance).toBe(1500);
    });

    it('他人は操作できない', () => {
      const account = new BankAccount('田中', 1000);
      const proxy = protectionProxy(account, '鈴木');

      expect(() => proxy.deposit(500)).toThrow('鈴木 は 田中 のアカウントを操作できません');
    });

    it('他人でも残高は確認できる', () => {
      const account = new BankAccount('田中', 1000);
      const proxy = protectionProxy(account, '鈴木');

      expect(proxy.balance).toBe(1000);
    });
  });

  describe('Virtual Proxy', () => {
    it('アクセスされるまでインスタンスが生成されない', () => {
      let created = false;
      const proxy = virtualProxy(() => {
        created = true;
        return new BankAccount('仮想', 5000);
      });

      expect(created).toBe(false);
      expect(proxy.balance).toBe(5000);
      expect(created).toBe(true);
    });

    it('一度生成されたインスタンスが再利用される', () => {
      let callCount = 0;
      const proxy = virtualProxy(() => {
        callCount++;
        return new BankAccount('仮想', 1000);
      });

      proxy.deposit(100);
      proxy.deposit(200);

      expect(callCount).toBe(1);
      expect(proxy.balance).toBe(1300);
    });
  });
});

File: apps/go/design-pattern/proxy/proxy_test.go
package proxy

import "testing"

func TestRealBankAccountDeposit(t *testing.T) {
	account := NewRealBankAccount(1000)
	if err := account.Deposit(500); err != nil {
		t.Fatalf("Deposit 失敗: %v", err)
	}
	if account.Balance() != 1500 {
		t.Errorf("期待値 1500, 実際 %d", account.Balance())
	}
}

func TestRealBankAccountWithdraw(t *testing.T) {
	account := NewRealBankAccount(1000)
	if err := account.Withdraw(300); err != nil {
		t.Fatalf("Withdraw 失敗: %v", err)
	}
	if account.Balance() != 700 {
		t.Errorf("期待値 700, 実際 %d", account.Balance())
	}
}

func TestRealBankAccountInsufficientFunds(t *testing.T) {
	account := NewRealBankAccount(100)
	err := account.Withdraw(200)
	if err == nil {
		t.Error("残高不足でエラーが返るべき")
	}
}

func TestProtectionProxyAllowsOwner(t *testing.T) {
	account := NewRealBankAccount(1000)
	proxy := NewProtectionProxy(account, "田中")

	if err := proxy.WithdrawAs("田中", 500); err != nil {
		t.Fatalf("所有者の出金が拒否された: %v", err)
	}
	if proxy.Balance() != 500 {
		t.Errorf("期待値 500, 実際 %d", proxy.Balance())
	}
}

func TestProtectionProxyBlocksNonOwner(t *testing.T) {
	account := NewRealBankAccount(1000)
	proxy := NewProtectionProxy(account, "田中")

	err := proxy.WithdrawAs("佐藤", 500)
	if err == nil {
		t.Error("非所有者の出金でエラーが返るべき")
	}
	if proxy.Balance() != 1000 {
		t.Errorf("残高が変わっていないべき: %d", proxy.Balance())
	}
}

func TestVirtualProxyDelaysCreation(t *testing.T) {
	vp := NewVirtualProxy(1000)
	if vp.IsInitialized() {
		t.Error("使用前に初期化されるべきではない")
	}
}

func TestVirtualProxyInitializesOnUse(t *testing.T) {
	vp := NewVirtualProxy(1000)
	_ = vp.Balance()
	if !vp.IsInitialized() {
		t.Error("使用後に初期化されるべき")
	}
}

func TestVirtualProxyDeposit(t *testing.T) {
	vp := NewVirtualProxy(1000)
	if err := vp.Deposit(500); err != nil {
		t.Fatalf("Deposit 失敗: %v", err)
	}
	if vp.Balance() != 1500 {
		t.Errorf("期待値 1500, 実際 %d", vp.Balance())
	}
}

func TestBankAccountInterface(t *testing.T) {
	// Compile-time interface satisfaction checks
	var _ BankAccount = &RealBankAccount{}
	var _ BankAccount = &ProtectionProxy{}
	var _ BankAccount = &VirtualProxy{}
}
